File: 2024/18.rs
use std::fs::File;
use std::io::{BufRead, BufReader};
use pathfinding::prelude::bfs;
mod helpers;
use helpers::in_bounds;

fn traverse(bytes: &Vec<(isize, isize)>, n_rows: usize, n_cols: usize, n: usize) -> Option<Vec<(isize, isize)>> {
    bfs(
        &(0, 0),
        |&(x, y)| {
            [(x + 1, y), (x, y + 1), (x - 1, y), (x, y - 1)]
                .into_iter()
                .filter({
                    move |&(i, j)| {
                        let index = bytes.iter().position(|&el| el == (i, j));
                        in_bounds(i, j, n_rows, n_cols) && (index.is_none() || index.unwrap() > n)
                    }
                })
        },
        |&_end| _end == (n_rows as isize - 1, n_cols as isize - 1),
    )
}

fn part_one(bytes: &Vec<(isize, isize)>, n_rows: usize, n_cols: usize) -> String {
    (traverse(&bytes, n_rows, n_cols, 1023).unwrap().len() - 1).to_string()
}

fn part_two(bytes: &Vec<(isize, isize)>, n_rows: usize, n_cols: usize) -> String {
    let mut lo: usize = 0;
    let mut hi: usize = bytes.len() - 1;
    while hi > lo + 1 {
        let mid: usize = (hi + lo) / 2;
        if traverse(&bytes, n_rows, n_cols, mid).is_some() {
            lo = mid;
        } else {
            hi = mid;
        }
    }
    format!("{},{}", bytes[hi].1, bytes[hi].0).to_string()
}

fn main_fn(res: &dyn Fn(&Vec<(isize, isize)>, usize, usize) -> String) -> String {
    let reader: BufReader<File> = BufReader::new(File::open("input").unwrap());
    let n_rows: usize = 71;
    let n_cols: usize = 71;
    let bytes: Vec<(isize, isize)> = reader
        .lines()
        .flatten()
        .map(|line| {
            let split: Vec<&str> = line.split(",").collect();
            (split[1].parse::<isize>().unwrap(), split[0].parse::<isize>().unwrap())
        })
        .collect();
    res(&bytes, n_rows, n_cols)
}

fn main() {
    println!("{}", main_fn(&part_one));
    println!("{}", main_fn(&part_two));
}

File: 2020/18.py
def part_one(q):
    res = q[0]
    i = 1
    while i < len(q):
        if q[i] == '+':
            res += q[i+1]
        elif q[i] == '*':
            res *= q[i+1]
        i += 2
    return res


def part_two(q):
    i = 0
    while i < len(q):
        if q[i] == '+':
            q[i] = q[i-1] + q[i+1]
            q.pop(i-1)
            q.pop(i)
            i -= 1
        i += 1
    res = q[0]
    i = 1
    while i < len(q):
        res *= q[i+1]
        i += 2
    return res


def create_req(string, i, fn):
    q = []
    while i < len(string):
        if string[i] in ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']:
            q.append(int(string[i]))
        elif string[i] in ['+', '*']:
            q.append(string[i])
        elif string[i] == '(':
            q_, i = create_req(string, i+1, fn)
            q.append(fn(q_))
        elif string[i] == ')':
            return q, i
        i += 1
    return q, i


def main(fn):
    total = 0
    with open("input", "r") as f:
        while True:
            string = f.readline().strip()
            if len(string) == 0:
                break
            q, i = create_req(string, 0, fn)
            total += fn(q)
    return total


if __name__ == "__main__":
    print(main(part_one))
    print(main(part_two))
